### PR TITLE
Refine HiveDBWriter

### LINKED ISSUE #685: Implement HiveDBWriter to load csv onto HDFS to improve the performance
**Is your feature request related to a problem? Please describe.**

SQLFlow can connect to the difference DB engine such as MySQL, hive, Maxcompute, SQLite3 and etc... Currently, SQLFLow Predict Job written the prediction result to the result table using [insert_values(driver, conn, table_name, table_schema, values)](https://github.com/sql-machine-learning/sqlflow/blob/develop/sql/python/sqlflow_submitter/db.py#L102)

But the *INSERT SQL* is ineffective for the most distributed DB engine. For example, each of the Hive INSERT SQL would launch a MapReduce Job on Yarn, so the Hive users usually load the local CSV file into HDFS which is the storage path of the Hive table.

**Describe the solution you'd like**

Implement a `DBWriter` interface in sqlflow_submitter package, the Python submitter program can using it like:
``` python
writer = db.db_writer_factory(driver="hive", conn=conn, table_name="prediciton_result", table_schema=["a", "b"], buff_size=100)
for ret in predciont_res:
    writer.write(ret)
writer.close()
```

We can implement `HiveDBWriter` like the code snippet:

``` python
class HiveDBWriter(DBWriter):
    ...
    def write(self, value):
        csv_file.write(value)

    def close(self):
	subprocess.call(["hdfs", "dfs", "-cp", "<local_csv_file>", "<hdfs_path>"])
```

TODO

- [x] Implement `DBWriter` interface. #686 
- [x] Load csv file into HDF file in the HiveDBWriter class

### PR DESCRIPTION
**Pull request by Yancey0623**
fixed #685

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -38,8 +38,8 @@ script:
   - docker build -t sqlflow:latest -f Dockerfile .
   - docker run --rm -v $GOPATH/src:/go/src -w /go/src/github.com/sql-machine-learning/sqlflow sqlflow:latest pre-commit run -a
   - docker run --rm -v $GOPATH:/go -w /go/src/github.com/sql-machine-learning/sqlflow sqlflow:latest bash scripts/test.sh
-  - docker run -d --name=hive -p 10000:10000 -p 10002:10002 -p 8040:8040 -p 8042:8042 -p 9864:9864 -p 9866:9866 -p 9867:9867 -p 9870:9870 -p 8020:8020 -p 8899:8899 sqlflow/gohive:dev python3 -m http.server 8899
-  - docker run --rm -v $GOPATH:/go --net=host --entrypoint bash -w /go/src/github.com/sql-machine-learning/sqlflow sqlflow:latest scripts/test_hive.sh
+  - docker run -d --name=hive sqlflow/gohive:dev python3 -m http.server 8899
+  - docker run --rm -v $GOPATH:/go --net=container:hive --entrypoint bash -w /go/src/github.com/sql-machine-learning/sqlflow sqlflow:latest scripts/test_hive.sh
```
Comment (typhoonzero) [inline .travis.yml:42]: Clever change !

File: scripts/image_build.sh [hunk 2]
```diff
@@ -115,3 +115,10 @@ curl https://raw.githubusercontent.com/sql-machine-learning/sqlflow/develop/exam
 
 # 9. install xgboost-launcher
 pip install xgboost-launcher==0.0.3
+
+# 10. install hadoop
```
Comment (typhoonzero) [inline scripts/image_build.sh]: install Hadoop to use as the client when writing CSV to hive tables.

File: sql/python/sqlflow_submitter/db_writer/hive.py [hunk 3]
```diff
@@ -11,22 +11,24 @@
...
+        tmp_f = tempfile.NamedTemporaryFile()
+        with open(tmp_f.name, "w") as f:
+            for row in self.rows:
+                line = "%s\n"  % '\001'.join([str(v) for v in row])
+                f.write(line)
+
+        hdfs_path = os.getenv("SQLFLOW_HIVE_LOCATION_ROOT_PATH", "/sqlflow")
+        namenode_addr = os.getenv("SQLFLOW_HDFS_NAMENODE_ADDR", "127.0.0.1:8020")
+        cmd_str = "hdfs dfs -copyFromLocal %s hdfs://%s%s/%s" % (f.name, namenode_addr, hdfs_path, self.table_name)
+        subprocess.run(cmd_str.split())
```
Comment (typhoonzero) [inline sql/python/sqlflow_submitter/db_writer/hive.py]: What if command run fails?

File: sql/python/sqlflow_submitter/db_writer/hive.py [hunk 4]
```diff
@@ -11,22 +11,24 @@
...
-        cursor = self.conn.cursor()
-        try:
-            cursor.executemany(statement, self.rows)
-            self.conn.commit()
-        finally:
-            cursor.close()
-            self.rows = []
+        tmp_f = tempfile.NamedTemporaryFile()
+        with open(tmp_f.name, "w") as f:
+            for row in self.rows:
```
Comment (typhoonzero) [inline sql/python/sqlflow_submitter/db_writer/hive.py]: seems this is to write rows in the buffer, is there a way to write all the rows before writing to hive?

File: sql/python/sqlflow_submitter/db_writer/hive.py [hunk 5]
```diff
@@ -11,22 +11,37 @@
...
+        for row in self.rows:
+            line = "%s\n"  % '\001'.join([str(v) for v in row])
+            self.f.write(line)
+        self.rows = []
+
+    def write_hive_table(self):
+        hdfs_path = os.getenv("SQLFLOW_HIVE_LOCATION_ROOT_PATH", "/sqlflow")
+        namenode_addr = os.getenv("SQLFLOW_HDFS_NAMENODE_ADDR", "127.0.0.1:8020")
+        cmd_str = "hdfs dfs -copyFromLocal %s hdfs://%s%s/%s" % (self.tmp_f.name, namenode_addr, hdfs_path, self.table_name)
+        subprocess.check_output(cmd_str.split())
```
Comment (typhoonzero) [inline sql/python/sqlflow_submitter/db_writer/hive.py:36]: can also add `stderr=subprocess.STDOUT` to output error message. Also need to print out the return output of the call `subprocess.check_output `

File: scripts/image_build.sh (hunk 2 above)
Comment (Yancey0623) [reply scripts/image_build.sh]: done.

**Review by typhoonzero — APPROVED** [review]

File: sql/python/sqlflow_submitter/db_writer/hive.py (hunk 5 above)
Comment (typhoonzero) [reply sql/python/sqlflow_submitter/db_writer/hive.py:36]: The error message seems will throw to stderr. So this may not needed.

**Pull request merged**